caddyhttp/httpserver/server.go

```diff
@@ -292,40 +285,99 @@ func (s *Server) Address() string {
 
 // Stop stops s gracefully (or forcefully after timeout) and
 // closes its listener.
-func (s *Server) Stop() (err error) {
-	s.Server.SetKeepAlivesEnabled(false)
-
-	if runtime.GOOS != "windows" {
-		// force connections to close after timeout
-		done := make(chan struct{})
-		go func() {
-			s.connWg.Done() // decrement our initial increment used as a barrier
-			s.connWg.Wait()
-			close(done)
-		}()
+func (s *Server) Stop() error {
+	if s.closed {
+		return errors.New("Server has been closed")
+	}
 
-		// Wait for remaining connections to finish or
-		// force them all to close after timeout
-		select {
-		case <-time.After(s.connTimeout):
-		case <-done:
+	// Make sure a listener was set
+	if s.listener != nil {
+		// Close the listener to stop all new connections
+		if err := s.listener.Close(); err != nil {
+			return err
 		}
 	}
 
-	// Close the listener now; this stops the server without delay
-	s.listenerMu.Lock()
-	if s.listener != nil {
-		err = s.listener.Close()
-		s.listener = nil
+	s.Server.SetKeepAlivesEnabled(false)
+	s.mu.Lock()
+	s.closed = true
+	s.mu.Unlock()
+
+	// force connections to close after timeout
+	wait := make(chan struct{})
+	go func() {
+		defer close(wait)
+		s.mu.Lock()
+		for c, st := range s.conns {
```

Comment:
@tpng Thinking something like this?

```
// Stop stops s gracefully by closing conns after s.connTimeout and
// closes its listener.
func (s *Server) Stop() error {
	if s.closed {
		return errors.New("Server has been closed")
	}

	// Make sure a listener was set
	if s.listener != nil {
		// Close the listener to stop all new connections
		if err := s.listener.Close(); err != nil {
			return err
		}
	}
	s.Server.SetKeepAlivesEnabled(false)

	// Give all connections time to finish. We wait the full timeout to close any
	// connection to avoid race situations with ConnState transitions
	time.Sleep(s.connTimeout)
	s.mu.Lock()
	s.closed = true
	for c, st := range s.conns {
		// Force close any idle and new connections.
		if st == http.StateIdle || st == http.StateNew {
			c.Close()
		}
	}
	s.mu.Unlock()

	// Closing this signals any TLS governor goroutines to exit
	if s.tlsGovChan != nil {
		close(s.tlsGovChan)
	}

	return nil
}
```

If we're always waiting for the s.connTimeout then I don't think we'd need the WaitGroup.